tests/test_qutrit_device.py

```diff
@@ -0,0 +1,1022 @@
+        state_probs = [0.1] * 9
+        state_probs[0] = 0.2
+
+        with pytest.raises(
+            qml.QuantumFunctionError,
+            match="The number of shots has to be explicitly set on the device",
+        ):
+            dev.sample_basis_states(number_of_states, state_probs)
+
+
+class TestStatesToTernary:
+    """Test the states_to_ternary method"""
+
+    def test_correct_conversion_two_states(self, mock_qutrit_device):
+        """Tests that the sample_basis_states method converts samples to binary correctly"""
+        wires = 4
+        shots = 10
+
+        number_of_states = 3**wires
+        basis_states = np.arange(number_of_states)
+        samples = np.random.choice(basis_states, shots)
+
+        dev = mock_qutrit_device()
+        res = dev.states_to_ternary(samples, wires)
+
+        expected = []
+
+        for s in samples:
+            num = []
+            for _ in range(wires):
+                num.append(s % 3)
+                s = s // 3
+
+            expected.append(num[::-1])
+
+        assert np.array_equal(res, np.array(expected))
+
+    test_ternary_conversion_data = [
+        (
+            np.array([2, 3, 2, 0, 0, 1, 6, 8, 5, 6]),
+            np.array(
+                [
+                    [0, 2],
+                    [1, 0],
+                    [0, 2],
+                    [0, 0],
+                    [0, 0],
+                    [0, 1],
+                    [2, 0],
+                    [2, 2],
+                    [1, 2],
+                    [2, 0],
+                ]
+            ),
+        ),
+        (
+            np.array([2, 7, 6, 8, 4, 1, 5]),
+            np.array(
+                [
+                    [0, 2],
+                    [2, 1],
+                    [2, 0],
+                    [2, 2],
+                    [1, 1],
+                    [0, 1],
+                    [1, 2],
+                ]
+            ),
+        ),
+        (
+            np.array([10, 7, 2, 15, 26, 20, 18, 24, 11, 6, 1, 0]),
+            np.array(
+                [
+                    [1, 0, 1],
+                    [0, 2, 1],
+                    [0, 0, 2],
+                    [1, 2, 0],
+                    [2, 2, 2],
+                    [2, 0, 2],
+                    [2, 0, 0],
+                    [2, 2, 0],
+                    [1, 0, 2],
+                    [0, 2, 0],
+                    [0, 0, 1],
+                    [0, 0, 0],
+                ]
+            ),
+        ),
+    ]
+
+    @pytest.mark.parametrize("samples, ternary_states", test_ternary_conversion_data)
+    def test_correct_conversion(self, mock_qutrit_device, samples, ternary_states, tol):
+        """Tests that the states_to_binary method converts samples to binary correctly"""
+        dev = mock_qutrit_device()
+        dev.shots = 5
+        wires = ternary_states.shape[1]
+        res = dev.states_to_ternary(samples, wires)
+        assert np.allclose(res, ternary_states, atol=tol, rtol=0)
+
+
```

Comment:
Note here that this is where the `TestExpval` and `TestVar` will be; there is also a `TestSample` in the qubit file, is that also coming afterwards or did I just miss it?

Reply:
I added `TestSample` yesterday. While I can currently get 100% code coverage for `QutritDevice.sample()`, more tests will be added to it in the future to test sampling observables that have eigenvalues. No such observables for qutrits have been added to PL yet, so I will leave a comment in the code saying so.